### PR TITLE
dtls13: additional epoch checks

### LINKED ISSUE #8579: [Bug]: DTLS 1.3 Server doesn't distinguish application_traffic_secret and handshake_traffic_secret
### Contact Details

_No response_

### Version

wolfssl-5.7.6

## Summary
I found that wolfssl 5.7.6 server didn't check if the messages are encrypted with correct traffic secret.

### Example 1: send Certificate, CertificateVerify, and Finished messages with epoch=3
Client should send Finished with epoch=2, which is using handshake_traffic_secret.
However, I deliberately send Finished with epoch=3 and encrypted with application_traffic_secret, we can still complete the handshake.

Image

### Example 2: [PSK] send Finished messages with epoch=3
the same problem as Example 1, but also exists when we using PSK instead of certificate authentication.

Image

### Example 3: send Application message with epoch=2
After we complete the handshake, we should use epoch=3 and encrypt our Application data with application_traffic_secret.
However, I intentionally use epoch=2 and encrypt Application data with handshake_traffic_secret, server still accept our Application data.

Image

## Reference
According to DTLS 1.3 RFC [rfc9147 2. Conventions and Terminology](https://www.rfc-editor.org/rfc/rfc9147.html#name-conventions-and-terminology)
> '{}'
> indicates messages protected using keys derived from a [sender]_handshake_traffic_secret.
> '[]'
> indicates messages protected using keys derived from traffic_secret_N.

and [rfc9147 6.1. Epoch Values and Rekeying](https://www.rfc-editor.org/rfc/rfc9147.html#dtls-msg-epoch) Figure 13:
```
Client                                             Server
------                                             ------

 Record 0
 ClientHello
 (epoch=0)
                            -------->
                                                     Record 0
                            <--------       HelloRetryRequest
                                                    (epoch=0)
 Record 1
 ClientHello                -------->
 (epoch=0)
                                                     Record 1
                            <--------             ServerHello
                                                    (epoch=0)
                                        {EncryptedExtensions}
                                                    (epoch=2)
                                                {Certificate}
                                                    (epoch=2)
                                          {CertificateVerify}
                                                    (epoch=2)
                                                   {Finished}
                                                    (epoch=2)
 Record 2
 {Certificate}              -------->
 (epoch=2)
 {CertificateVerify}
 (epoch=2)
 {Finished}
 (epoch=2)
                                                     Record 2
                            <--------                   [ACK]
                                                    (epoch=3)
 Record 3
 [Application Data]         -------->
 (epoch=3)
                                                     Record 3
                            <--------      [Application Data]
                                                    (epoch=3)
```

---
### Environment
OS: Ubuntu 24.04

### How I build wolfSSL:
```bash
bash autogen.sh
AM_CFLAGS='-DHAVE_AES_CBC -DWOLFSSL_AES_128 -DWOLFSSL_DEBUG_TLS' ./configure --enable-dtls --enable-dtls13 --enable-keylog-export --enable-psk --enable-rsa --enable-sha --enable-debug C_EXTRA_FLAGS=-DWOLFSSL_STATIC_PSK
```
### How I launch wolfSSL server:
```bash
# PSK authentication
/dtls-fuzzer/suts/wolfssl-5.7.6/examples/server/server  -i -x -p 17830 -s -u -v 4 -l TLS13-AES128-GCM-SHA256

# Certificate authentication with Client certificate required
/dtls-fuzzer/suts/wolfssl-5.7.6/examples/server/server  -i -x -p 17680 -u -v 4 -l TLS13-AES128-GCM-SHA256 -c /dtls-fuzzer/experiments/keystore/rsa2048_cert.pem -k /dtls-fuzzer/experiments/keystore/rsa2048_key.pem -A /dtls-fuzzer/experiments/keystore/rsa2048_cert.pem
```

### Reproduction steps

If you don't want to do reproduction, then you can download **reproduction.zip**, I collected network traffic captures, sslkeylog files, wolfssl server debug logs, and Docker files inside it.
[reproduction.zip](https://github.com/user-attachments/files/19391005/reproduction.zip)

---

In order to make the reproduction easy to run, I created [docker images](https://hub.docker.com/r/west2077/wolfssl-bug-reproduction/tags), so you can generate results with only one command.
If you are using Linux, you need to install [Docker](https://docs.docker.com/engine/install/ubuntu/) first.
If you are using Windows, you need to install [Docker Desktop](https://www.docker.com/products/docker-desktop/) first.

After you run the command, you can see a new folder `myTmpFolder`, inside the folder there will be three files:
* dtls_capture: network traffic captured by tcpdump, and you can check it with Wireshark
* running_output: Java program logs and wolfssl server debug logs
* sslkeylog.log: generated by wolfssl, you can use it to decrypt network traffic within Wireshark

### For example 1: send Certificate, CertificateVerify, and Finished messages with epoch=3
Linux:
`docker run --rm -v $(pwd)/myTmpFolder:/output west2077/wolfssl-bug-reproduction:finished-with-epoch3`
Windows with powershell:
`docker run --rm -v ${PWD}/myTmpFolder:/output west2077/wolfssl-bug-reproduction:finished-with-epoch3`
### For example 2: [PSK] send Finished messages with epoch=3
Linux: 
`docker run --rm -v $(pwd)/myTmpFolder:/output west2077/wolfssl-bug-reproduction:finished-with-epoch3-psk`
Windows with powershell: 
`docker run -it --rm -v ${PWD}/myTmpFolder:/output west2077/wolfssl-bug-reproduction:finished-with-epoch3-psk`

### For example 3: send Application message with epoch=2
Linux: 
`docker run --rm -v $(pwd)/myTmpFolder:/output west2077/wolfssl-bug-reproduction:application-with-epoch2`
Windows with powershell: 
`docker run --rm -v ${PWD}/myTmpFolder:/output west2077/wolfssl-bug-reproduction:application-with-epoch2`

## Dockerfile and How I build the docker image
I choose Docker for easy to run, portability, and stable reproduction, and you can also build the docker image from scratch.
The only file you need is Dockerfile [Dockerfile_finished_with_epoch3](https://github.com/c-southwest/dtls-fuzzer/blob/add-DTLS-13-wolfssl-bug-reproduction/Dockerfile_finished_with_epoch3), [Dockerfile_finished_with_epoch3_psk](https://github.com/c-southwest/dtls-fuzzer/blob/add-DTLS-13-wolfssl-bug-reproduction/Dockerfile_finished_with_epoch3_psk), or [Dockerfile_application_with_epoch2](https://github.com/c-southwest/dtls-fuzzer/blob/add-DTLS-13-wolfssl-bug-reproduction/Dockerfile_application_with_epoch2).
After you download the Dockerfile, you can build the image with command:
### For example 1: send Certificate, CertificateVerify, and Finished messages with epoch=3
`docker build -t wolfssl-bug-reproduction:finished-with-epoch3 -f Dockerfile_finished_with_epoch3 .`

### For example 2: [PSK] send Finished messages with epoch=3
`docker build -t wolfssl-bug-reproduction:finished-with-epoch3-psk -f Dockerfile_finished_with_epoch3_psk .`

### For example 3: send Application message with epoch=2
`docker build -t wolfssl-bug-reproduction:application-with-epoch2 -f Dockerfile_application_with_epoch2 .`

### Relevant log output

```shell

```

### LINKED ISSUE #8627: [Bug]: DTLS 1.3 Client doesn't reject Application message encrypted with handshake_traffic_secret
### Contact Details

GitHub should be enough

### Version

wolfssl-5.7.6

### Description

This issue is similar to #8579 , the difference is that the problem "Application with epoch=2" also exists in wolfssl **Client** implementation. However, wolfssl **Client** implementation doesn't have the problem "Finished with epoch=3".

Image

#### How I build wolfSSL:
```bash
bash autogen.sh
AM_CFLAGS='-DHAVE_AES_CBC -DWOLFSSL_AES_128 -DWOLFSSL_DEBUG_TLS' ./configure --enable-dtls --enable-dtls13 --enable-keylog-export --enable-psk --enable-rsa --enable-sha --enable-debug C_EXTRA_FLAGS=-DWOLFSSL_STATIC_PSK
```
#### How I launch wolfSSL client:
```bash
# PSK authentication
/home/west/dtls-fuzzer/suts/wolfssl-5.7.6/examples/client/client  -i -x -p 28092 -s -u -v 4 -l TLS13-AES128-GCM-SHA256
```

### Reproduction steps

Before sending Application message, I deliberately set write epoch to 2, so as a Server I can send Application with epoch=2.

You can find Wireshark capture, sslkeylog.log, and also wolfssl client debug logs.
[captures_and_logs.zip](https://github.com/user-attachments/files/19573957/captures_and_logs.zip)

### Relevant log output

```shell
wolfSSL Leaving wolfSSL_dtls_get_current_timeout, return 320
wolfSSL Entering EmbedReceiveFrom
wolfSSL Entering wolfSSL_dtls_get_using_nonblock
Data received
	2e 39 1e 00 1f 68 80 24 1b bd 51 00 f4 fb 2f 61 |.9...h.$..Q.../a
	9b 71 ee c4 88 07 82 da 07 79 8e ef c9 4d bb 5f |.q.......y...M._
	61 31 66 50                                     |a1fP
reconstructed epoch number: 2
wolfSSL Entering Dtls13SetEpochKeys
Provisioning DECRYPT key
	bf 08 cb 18 02 6c c4 fb 62 8a 17 77 ee f8 5a 71 |.....l..b..w..Zq
Provisioning AES Record Number dec key:
	e3 13 7a ae 10 93 3d d1 1a a3 f7 d3 ed f3 89 7d |..z...=........}
reconstructed seq number: 2
wolfSSL Entering DecryptTls13
Data to decrypt
	68 80 24 1b bd 51 00 f4 fb 2f 61 9b 71 ee c4    |h.$..Q.../a.q..
Additional Authentication Data
	2e 00 02 00 1f                                  |.....
Authentication tag
	88 07 82 da 07 79 8e ef c9 4d bb 5f 61 31 66 50 |.....y...M._a1fP
Dtls13GetSeq(): using seq: 2
TLS 1.3 Write Sequence 0 2
Nonce
	92 14 25 3c ff bb 3b 04 95 dc f8 2d             |..%<..;....-
Decrypted data
	68 65 6c 6c 6f 20 77 6f 6c 66 73 73 6c 21 17    |hello wolfssl!.
wolfSSL Entering Dtls13CheckWindow
wolfSSL Entering Dtls13UpdateWindow
received record layer msg
got app DATA
wolfSSL Entering Dtls13DoScheduledWork
Shrinking input buffer
wolfSSL Leaving ReceiveData(), return 14
wolfSSL Leaving wolfSSL_read_internal, return 14
```

### PR DESCRIPTION
**Pull request by julek-wolfssl**

### CODE REVIEW

File: src/dtls13.c [hunk 1]
```diff
@@ -1640,6 +1640,102 @@ static int Dtls13AcceptFragmented(WOLFSSL *ssl, enum HandShakeType type)
...
+            case client_key_exchange:
+                if (!w64Equal(ssl->keys.curEpoch64, hsEpoch)) {
+                    if (ssl->options.side == WOLFSSL_CLIENT_END &&
+                            ssl->options.serverState < SERVER_HELLO_COMPLETE) {
+                        /* before processing SH we don't know which version
+                         * will be negotiated.  */
+                        if (!w64Equal(ssl->keys.curEpoch64, plainEpoch)) {
+                            WOLFSSL_MSG("Msg should be epoch 2 or 0");
+                            WOLFSSL_ERROR_VERBOSE(SANITY_MSG_E);
+                            return SANITY_MSG_E;
```
Comment (rizlik) [inline src/dtls13.c:1675]: As message processing happens in order, we already negotiated the version on this message.
We can return error on 1.2 only messages (is it already checked?) or directly check that the messages are in HANDSHAKE_EPOCH

File: src/dtls13.c [hunk 2]
```diff
@@ -1640,6 +1640,102 @@ static int Dtls13AcceptFragmented(WOLFSSL *ssl, enum HandShakeType type)
...
+                        }
+                    }
+                }
+                else {
+                    /* Allow epoch 2 in case of rtx */
+                    if (!w64GTE(ssl->keys.curEpoch64, hsEpoch)) {
+                        WOLFSSL_MSG("Msg should be epoch 2+");
+                        WOLFSSL_ERROR_VERBOSE(SANITY_MSG_E);
+                        return SANITY_MSG_E;
```
Comment (rizlik) [inline src/dtls13.c:1713]: we can be more strictly with post-auth hs messages.
once we decrypt an hs message with seq X in epoch Y we can error out on every message with message seq N <= X in epoch Z <= Y

File: src/dtls13.c [hunk 3]
```diff
@@ -1640,6 +1640,102 @@ static int Dtls13AcceptFragmented(WOLFSSL *ssl, enum HandShakeType type)
...
+                }
+                break;
+            case certificate_status:
+            case change_cipher_hs:
+            case key_update:
+            case session_ticket:
+                if (!w64GTE(ssl->keys.curEpoch64, t0Epoch)) {
+                    WOLFSSL_MSG("Msg should be epoch 3+");
+                    WOLFSSL_ERROR_VERBOSE(SANITY_MSG_E);
+                    return SANITY_MSG_E;
```
Comment (rizlik) [inline src/dtls13.c:1724]: similar as above probably

File: src/dtls13.c [hunk 4]
```diff
@@ -1640,6 +1640,102 @@ static int Dtls13AcceptFragmented(WOLFSSL *ssl, enum HandShakeType type)
...
+                }
+                break;
+            case end_of_early_data:
+            case message_hash:
+            case no_shake:
+            default:
+                WOLFSSL_MSG("Unknown message type");
+                WOLFSSL_ERROR_VERBOSE(SANITY_MSG_E);
+                return SANITY_MSG_E;
```
Comment (rizlik) [inline src/dtls13.c:1733]: I don't think we can get here with 1.2 only messages.

File: tests/api/test_dtls.c [hunk 5]
```diff
@@ -595,3 +595,54 @@ int test_wolfSSL_dtls_cid_parse(void)
 #endif
     return EXPECT_RESULT();
 }
+
+int test_dtls13_epochs(void) {
```
Comment (rizlik) [inline tests/api/test_dtls.c:599]: Nice to Have: a test with a complete handshake and/or a test with DTLS fuzzer.

**Review by rizlik — CHANGES_REQUESTED** [review]

File: src/dtls13.c (hunk 2 above)
Comment (julek-wolfssl) [reply src/dtls13.c:1713]: I think it would be hard to handle edge cases for when each msg should arrive. This check is a sanity check that they arrive in a legal epoch.

File: src/dtls13.c (hunk 4 above)
Comment (julek-wolfssl) [reply src/dtls13.c:1733]: This is handling of all possible values of `HandShakeType`.

File: tests/api/test_dtls.c (hunk 5 above)
Comment (julek-wolfssl) [reply tests/api/test_dtls.c:599]: Feasible once we have a reliable framework for generating test messages directly. We can consider this for next major TLS improvements.

File: src/dtls13.c (hunk 1 above)
Comment (rizlik) [reply src/dtls13.c:1671]: I see what you mean, but does this allow plaintext encrypted extension in 1.3 if they arrive out of order? In that case I think we should check that as well

Comment (julek-wolfssl) [reply src/dtls13.c:1671]: Encryption status check is handled in `MsgCheckEncryption`.

**Comment by julek-wolfssl** [comment]
AgentOfflineException

**Review by rizlik — DISMISSED** [review]

**Comment by dgarske** [comment]
Retest this please. History lost

File: wolfssl/internal.h [hunk 6]
```diff
@@ -2222,7 +2222,7 @@ WOLFSSL_LOCAL int DoFinished(WOLFSSL* ssl, const byte* input, word32* inOutIdx,
 WOLFSSL_LOCAL int DoTls13Finished(WOLFSSL* ssl, const byte* input, word32* inOutIdx,
                            word32 size, word32 totalSz, int sniff);
 #endif
-WOLFSSL_LOCAL int DoApplicationData(WOLFSSL* ssl, byte* input, word32* inOutIdx,
+WOLFSSL_API int DoApplicationData(WOLFSSL* ssl, byte* input, word32* inOutIdx,
```
Comment (douzzer) [inline wolfssl/internal.h:2225]: Shouldn't this be `WOLFSSL_TEST_VIS`?

We don't have any `WOLFSSL_API`s in `internal.h` anymore.

File: wolfssl/internal.h [hunk 7]
```diff
@@ -7052,6 +7052,7 @@ WOLFSSL_LOCAL int Dtls13HandshakeSend(WOLFSSL* ssl, byte* output,
     word16 output_size, word16 length, enum HandShakeType handshake_type,
     int hash_output);
 WOLFSSL_LOCAL int Dtls13RecordRecvd(WOLFSSL* ssl);
+WOLFSSL_API int Dtls13CheckEpoch(WOLFSSL* ssl, enum HandShakeType type);
```
Comment (douzzer) [inline wolfssl/internal.h:7055]: Same -- either `WOLFSSL_TEST_VIS`, or it doesn't belong in `internal.h`.

**Review by douzzer — CHANGES_REQUESTED** [review]
looks good, but visibility attributes need fixes in `internal.h`.

File: wolfssl/internal.h (hunk 6 above)
Comment (julek-wolfssl) [reply wolfssl/internal.h]: fixed

File: wolfssl/internal.h (hunk 7 above)
Comment (julek-wolfssl) [reply wolfssl/internal.h]: fixed

**Review by dgarske — APPROVED** [review]

**Review by douzzer — APPROVED** [review]
